docs/gen-ai/llm-spans.md

```diff
@@ -79,52 +96,473 @@ These attributes track input data and metadata for a request to an LLM. Each att
 
 ## Events
 
-In the lifetime of an LLM span, an event for prompts sent and completions received MAY be created, depending on the configuration of the instrumentation.
+In the lifetime of an GenAI span, an event for each message application sends to GenAI and receives in response from it MAY be created, depending on the configuration of the instrumentation.
+The generic events applicable to multiple GenAI models follow `gen_ai.{type}.message` naming pattern.
+GenAI vendor-specific instrumentations SHOULD follow `gen_ai.{gen_ai.system}.{type}.message` pattern to record system-specific events.
+
+It's RECOMMENDED to use [Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md) to record GenAI events once it's implemented in corresponding language.
```

Comment:
@codefromthecrypt it's recommended in part because events (i.e., named logs) are sent as a separate batch from trace data. Many telemetry backends have harder restrictions on trace data, mostly because of attribute count and size limitations, but they typically do not have anywhere near those kinds of restrictions on log data. In a perfect world, this data would probably be captured as a part of a trace tracking the call/response to an LLM, but tracing systems at large are not design to handle large payloads.[0]

That said, the events API is very new, and it's not likely that a given implementation language will have events implemented. It would be expected that a given language that logs this data as a part of a trace (or something else) eventually moves to the events API once it's available for them.

[0] There are several exceptions to this on the market today, but it's not generally the case, and even these exceptions have limits you can run into

Reply:
@cartermp sorry for sidetracking as I'm sure folks have thought about this, but could we establish an event bridge instead of special casing here? at least then it would be easier to reason with and then a mechanical concern vs special to this LLM.

I wrote a spec for wasm which had similar feature concerns. e.g. is buffering available? basically if we know that events aren't available and it doesn't matter the reason albeit lang or backend, then we have a compensating alternative. Some may choose to not enable that alternative.

ack feel free to punt me.

Reply:
> but could we establish an event bridge instead of special casing here?

Can you clarify what you mean by this? In OTel, logs are already defined as a part of the "logs bridge" concept, and events are just logs with a name. But this may be a different terminology from what you're referring to.

Reply:
Sorry got very sick and was offline.

So for the highlighted line, we are saying to use events when they are implemented:

> It's RECOMMENDED to use [Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md) to record GenAI events once it's implemented in corresponding language.

I meant to ask if there was (or could be) a canonical way to adapt a spec aimed for the Event API until such time as it is available. If there was a way (call it X)..

> It's RECOMMENDED to use [Event API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/event-api.md) to record GenAI events. Instrumentation can use X to adapt to the Event API in languages or integrations where it is not yet available.

What had me mostly on this line is the inability to link to a resolution of the event api being unavailable. If an interim approach X was linked, it could be updated later saying "this is deprecated as all languages now have the event API" and those looking at this doc could navigate to that conclusion.

It is about the stale doc problem for a concern that appears temporary. Hope this helps.

Reply:
you can close this comment. I can follow-up later on which languages support log events API. It isn't blocking feedback.

Reply:
There is a matrix here: https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md#events (short answer - none, but it's in development), it sometimes possible to abuse logs api to send propoer log-based events

Reply:
I'll add  link to the text

Reply:
done, closing for now